subxt/src/rpc/lightclient/background.rs

```diff
@@ -0,0 +1,425 @@
+                        // Track this subscription ID if send is successful.
+                        self.subscriptions.insert(sub_id, sender);
+                    }
+                }
+            }
+            Ok(RpcResponse::Subscription { method, id, result }) => {
+                // Subxt calls into `author_submitAndWatchExtrinsic`, however the smoldot produces
+                // `{"event":"broadcasted","numPeers":1}` and `{"event":"validated"}` which are part
+                // of the RPC V2 API. Ignore those events.
+                if method == "transaction_unstable_watchEvent"
+                    && (result.to_string().contains("broadcasted")
+                        || result.to_string().contains("validated"))
+                {
+                    tracing::debug!(target: LOG_TARGET, "Ignoring notification {:?}", result);
+                    return;
+                }
+
+                if let Some(sender) = self.subscriptions.get_mut(&id) {
+                    // Send the current notification response.
+                    if sender.send(result).await.is_err() {
+                        tracing::warn!(
+                            target: LOG_TARGET,
+                            "Cannot send notification to subscription {:?}",
+                            id
+                        );
+
+                        // Remove the sender if the subscription dropped the receiver.
+                        self.subscriptions.remove(&id);
+                    }
+                }
+            }
+            Err(err) => {
+                tracing::warn!(target: LOG_TARGET, "cannot decode RPC response {:?}", err);
+            }
+        }
+    }
+
+    /// Perform the main background task:
+    /// - receiving user registration for RPC method / subscriptions
+    /// - providing the results from the light client back to users.
+    pub async fn start_task(
+        &mut self,
+        backend: mpsc::Receiver<BackendMessage>,
+        rpc_responses: smoldot_light::JsonRpcResponses,
+    ) {
+        let backend_event = tokio_stream::wrappers::ReceiverStream::new(backend);
+        let rpc_responses_event =
+            futures_util::stream::unfold(rpc_responses, |mut rpc_responses| async {
+                rpc_responses
+                    .next()
+                    .await
+                    .map(|result| (result, rpc_responses))
+            });
+
+        tokio::pin!(backend_event, rpc_responses_event);
+
+        let mut backend_event_fut = backend_event.next();
+        let mut rpc_responses_fut = rpc_responses_event.next();
+
+        loop {
+            match future::select(backend_event_fut, rpc_responses_fut).await {
+                // Message received from the backend: user registered.
+                Either::Left((backend_value, previous_fut)) => {
+                    let Some(message) = backend_value else {
+                        tracing::trace!(target: LOG_TARGET, "Frontend channel closed");
```

Comment:
This is exactly one of the confusing things; the thing is called `backend_value` but if it's not there, the "frontend" is closed :D